app/Services/ProvisioningService.php

```diff
@@ -0,0 +1,366 @@
+<?php
+
+namespace App\Services;
+
+use App\Enums\ServerStatus;
+use App\Enums\TimePeriod;
+use App\Models\Permission;
+use App\Models\Role;
+use App\Models\RoomType;
+use App\Models\Server;
+use App\Models\ServerPool;
+use App\Models\User;
+use App\Settings\GeneralSettings;
+use App\Settings\RecordingSettings;
+use App\Settings\RoomSettings;
+use App\Settings\UserSettings;
+use Illuminate\Database\RecordsNotFoundException;
+use Illuminate\Support\Facades\Validator;
+use Illuminate\Validation\Rule;
+use Log;
+use ReflectionClass;
+use UnexpectedValueException;
+use ValueError;
+
+abstract class AbstractProvisioner
+{
+    protected string $model;
+
+    protected array $expectedProperties;
+
+    protected function instanceName(object $properties)
+    {
+        return $properties->name;
+    }
+
+    protected function modelName()
```

Comment:
This could be computed in the constructor and saved as a variable, instead of beeing computed multiple times.
Maybe make this and the variable private and change $model to a const

Reply:
I can't wrap my head around PHP's concept of constants but I'm using constructors and private vars to configure concrete provisioners now.